File: 1370-count-number-of-nice-subarrays/solution.java
class Solution {
    public int numberOfSubarrays(int[] nums, int k) {
        return atmost(nums, k) - atmost(nums, k-1);
    }

    int atmost(int[] nums, int k) {
        int i=0;
        int j=0;
        int oddCount = 0;
        int ans = 0;
        int n = nums.length;

        while(j < n) {
            if(nums[j]%2 == 1) {
                oddCount++;
            }

            while(oddCount > k) {
                // shrink the window
                if(nums[i]%2 == 1) {
                    oddCount--;
                }
                i++;
            }
            ans += j-i+1;
            j++;
        }
        return ans;
    }
}

File: 1370-count-number-of-nice-subarrays/solution.swift
class Solution {
    func numberOfSubarrays(_ nums: [Int], _ k: Int) -> Int {
        var n = nums.count

        func atmostk(_ k: Int) -> Int {
            var i=0
            var j=0
            var oddCount = 0
            var count = 0

            while j < n {
                if nums[j]%2 == 1 {
                    oddCount += 1 //3
                }
                
                // invalid - shrink
                while oddCount > k {
                    if nums[i]%2 == 1 {
                        oddCount -= 1
                    }
                    i += 1//1
                }
                count += j-i+1//1
                j += 1//3
            }
            return count
        }
    return atmostk(k) - atmostk(k-1)
    }
}
